src/main/java/org/venice/beachfront/bfapi/BfApiConfig.java

```diff
@@ -92,4 +98,183 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
+					.setMaxConnTotal(httpMaxTotal)
+					.setMaxConnPerRoute(httpMaxRoute)
+					.setSSLHostnameVerifier(new NoopHostnameVerifier())
+					.setKeepAliveStrategy(getKeepAliveStrategy())
+					.build();
+
+			restTemplate.setRequestFactory(new HttpComponentsClientHttpRequestFactory(httpClient));
+
+			return restTemplate;
+		}
+
+		@Bean
+		public AuthHeaders authHeaders() {
+			return new BasicAuthHeaders();
+		}
+	}
+
+	@Configuration
+	@Profile({ "pki-geoserver-auth" })
+	protected static class PKIAuthenticationConfig {
+
+		@Value("${http.max.total}")
+		private int httpMaxTotal;
+
+		@Value("${http.max.route}")
+		private int httpMaxRoute;
+
+		@Value("${JKS_FILE}")
+		private String keystoreFileName;
+
+		@Value("${JKS_PASSPHRASE}")
+		private String keystorePassphrase;
+
+		@Value("${PZ_PASSPHRASE}")
+		private String piazzaKeyPassphrase;
+
+		@Bean
+		public RestTemplate restTemplate() throws KeyManagementException, UnrecoverableKeyException,
+				NoSuchAlgorithmException, KeyStoreException, CertificateException, IOException {
+
+			final SSLContext sslContext = SSLContexts.custom()
+					.loadKeyMaterial(getStore(), piazzaKeyPassphrase.toCharArray())
+					.loadTrustMaterial(null, new TrustSelfSignedStrategy())
+					.useProtocol("TLS")
+					.build();
+
+			final Registry<CookieSpecProvider> registry = RegistryBuilder.<CookieSpecProvider>create()
+					.register("myspec", new MySpecProvider())
+					.build();
+
+			final RequestConfig requestConfig = RequestConfig.custom()
+					.setCookieSpec("myspec")
+					.setCircularRedirectsAllowed(true)
+					.build();
+
+			final HttpClient httpClient = HttpClientBuilder.create()
+					.setDefaultRequestConfig(requestConfig)
+					.setMaxConnTotal(httpMaxTotal)
+					.setSSLContext(sslContext)
+					.setSSLHostnameVerifier(new NoopHostnameVerifier())
+					.setDefaultCookieStore(new BasicCookieStore())
+					.setDefaultCookieSpecRegistry(registry)
+					.setRedirectStrategy(new MyRedirectStrategy())
+					.setMaxConnPerRoute(httpMaxRoute)
+					.setKeepAliveStrategy(getKeepAliveStrategy())
+					.build();
+
+			final RestTemplate restTemplate = new RestTemplate();
+			restTemplate.setRequestFactory(new HttpComponentsClientHttpRequestFactory(httpClient));
+
+			final List<HttpMessageConverter<?>> messageConverters = new ArrayList<HttpMessageConverter<?>>();
+			messageConverters.add(new StringHttpMessageConverter());
+			messageConverters.add(new MappingJackson2HttpMessageConverter());
+			restTemplate.setMessageConverters(messageConverters);
+
+			return restTemplate;
+		}
+
+		@Bean
+		public AuthHeaders authHeaders() {
+			return new PKIAuthHeaders();
+		}
+
+		protected KeyStore getStore()
+				throws KeyStoreException, IOException, CertificateException, NoSuchAlgorithmException {
+			final KeyStore store = KeyStore.getInstance(KeyStore.getDefaultType());
+			final InputStream inputStream = getClass().getClassLoader().getResourceAsStream(keystoreFileName);
+			try {
```

Comment:
This works as-is but you could do a try-with-resources here and take advantage of auto-closing streams.

Again, not necessary, but a cool feature Java has now.

Reply:
Agreed, Done.